tests/integration/replication-4.tcl

```diff
@@ -85,6 +85,42 @@ start_server {tags {"repl external:skip"}} {
     }
 }
 
+start_server {tags {"repl external:skip"}} {
+    start_server {} {
+        set master [srv -1 client]
+        set master_host [srv -1 host]
+        set master_port [srv -1 port]
+        set slave [srv 0 client]
+
+        test {First server should have role slave after SLAVEOF} {
+            $slave slaveof $master_host $master_port
+            wait_for_condition 50 100 {
+                [s 0 master_link_status] eq {up}
+            } else {
+                fail "Replication not started."
+            }
+        }
+
+        test {Replication of an expired key does not delete the expired key} {
+            $master debug set-active-expire 0
+            $master set k 1 ex 1
+            exec kill -SIGSTOP [srv 0 pid]
+            $master incr k
+            after 1000
+            exec kill -SIGCONT [srv 0 pid]
+            wait_for_condition 50 100 {
+                [status $slave slave_repl_offset] ==
+                [status $master master_repl_offset]
+            } else {
+                fail "Replication did not catch up."
+            }
+            # Check that k is expired but still exists in the replica.
+            assert_equal 0 [$slave exists k]
+            $slave debug object k ; # Raises exception if k does not exist.
```

Comment:
i think the propagation of INCR (with the bug) would still mean the key exists.
i.e. it would be deleted as expired, and it would create a new non-volatile key with the value of 1.
so what i think we should do to verify is disable active expiry on the replica too.
then promote it, and use TTL and GET to check the value.

Reply:
The line above checks that `exists k` returns 0, which means that the key is expired. Without the fix in expireIfNeeded, exists returns 1 here.

Reply:
... because if we have a non-volatile key at the end, the line `assert_equal 0 [$slave exists k]` would fail.

Reply:
Without the check `server.current_client == server.master` in expireIfNeeded, the sequence is like you described and but then the assert fails because `exists k` returns `1`.

Reply:
Ohh, OK, not because the key is volatile vs non-volatile, but because we know that if it's volatile, it's also already logically expired. 
And because active expiry is disabled and there are no other activity on the master so it shouldn't send a DEL, so we can trust that.

Reply:
Exactly. Focus on "expired", not on "volatile". I avoided the word "exists" in the comments because logically it doesn't exist. But it's still there, in the dict in the replica.